{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### \n",
    "#### SPARCLE - code for assigning transcripts to cells using MLE\n",
    "#### Code version 1\n",
    "####\n",
    "##### Code author - Sandhya Prabhakaran\n",
    "#### "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "import numpy as np\n",
    "np.set_printoptions(threshold=sys.maxsize)\n",
    "\n",
    "import time\n",
    "\n",
    "from matplotlib.patches import Polygon\n",
    "\n",
    "from PIL import Image, ImageDraw\n",
    "from skimage import measure #package is scikit-image\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "plt.close('all')\n",
    "\n",
    "from scipy.spatial import distance\n",
    "\n",
    "import phenograph as pg\n",
    "import cv2\n",
    "\n",
    "from sklearn.mixture import BayesianGaussianMixture\n",
    "from sklearn.manifold import TSNE\n",
    "\n",
    "#import matplotlib\n",
    "#matplotlib.use('Agg')\n",
    "\n",
    "import random\n",
    "from scipy.stats import zscore\n",
    "\n",
    "import scipy.stats as stats\n",
    "from scipy.stats import multivariate_normal\n",
    "\n",
    "from joblib import Parallel, delayed\n",
    "import multiprocessing\n",
    "\n",
    "import scipy.io\n",
    "from scipy.io import mmread\n",
    "\n",
    "import colorsys\n",
    "\n",
    "import seaborn as sns\n",
    "import fastcluster\n",
    "\n",
    "\n",
    "from scipy.spatial import distance\n",
    "\n",
    "from numpy.lib.stride_tricks import as_strided\n",
    "\n",
    "import tifffile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "############\n",
    "#### update path variables for data\n",
    "############\n",
    "\n",
    "\n",
    "path_data = os.getcwd() \n",
    "path_t = os.getcwd() \n",
    "path_python = path_t\n",
    "\n",
    "path_scseq = os.path.join(path_data + '/GSE113576_matrix.mtx')\n",
    "path_cb = os.path.join(path_data + '/M22E1_codebook.csv')\n",
    "path_temp_2 = os.path.join(path_data + '/GSE113576_genes.tsv')\n",
    "\n",
    "path_Allen = os.path.join(path_data+'/SongLinROIS_deduplicated.json')\n",
    "brian_csv_path = os.path.join(path_data+'/smFISH_MCT_CZI_Panel_0_spot_table.csv')\n",
    "\n",
    "#read in the cluster ids\n",
    "path_sc_cellid = os.path.join(path_data + '/aau5324_Moffitt_Table-S1.xlsx')\n",
    "\n",
    "fov_count = 'current_run/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### D.Bartel: https://github.com/danbar/qr_decomposition\n",
    "\n",
    "\"\"\"Module for Givens rotation.\"\"\"\n",
    "\n",
    "from math import copysign, hypot\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def givens_rotation(A):\n",
    "    \"\"\"Perform QR decomposition of matrix A using Givens rotation.\"\"\"\n",
    "    (num_rows, num_cols) = np.shape(A)\n",
    "\n",
    "    # Initialize orthogonal matrix Q and upper triangular matrix R.\n",
    "    Q = np.identity(num_rows)\n",
    "    R = np.copy(A)\n",
    "\n",
    "    # Iterate over lower triangular matrix.\n",
    "    (rows, cols) = np.tril_indices(num_rows, -1, num_cols)\n",
    "    for (row, col) in zip(rows, cols):\n",
    "\n",
    "        # Compute Givens rotation matrix and\n",
    "        # zero-out lower triangular matrix entries.\n",
    "        if R[row, col] != 0:\n",
    "            (c, s) = _givens_rotation_matrix_entries(R[col, col], R[row, col])\n",
    "\n",
    "            G = np.identity(num_rows)\n",
    "            G[[col, row], [col, row]] = c\n",
    "            G[row, col] = s\n",
    "            G[col, row] = -s\n",
    "\n",
    "            R = np.dot(G, R)\n",
    "            Q = np.dot(Q, G.T)\n",
    "\n",
    "    return (Q, R)\n",
    "\n",
    "\n",
    "def _givens_rotation_matrix_entries(a, b):\n",
    "    \"\"\"Compute matrix entries for Givens rotation.\"\"\"\n",
    "    r = hypot(a, b)\n",
    "    c = a/r\n",
    "    s = -b/r\n",
    "\n",
    "    return (c, s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def connectpoints(x,y,p1,p2):\n",
    "    x1, x2 = x[p1], x[p2]\n",
    "    y1, y2 = y[p1], y[p2]\n",
    "    plt.plot([x1,x2],[y1,y2],'k-',linewidth=0.5)\n",
    "    \n",
    "    \n",
    "def setdiff2d(lst1,lst2):\n",
    "    aset = set([tuple(x) for x in lst1])\n",
    "    bset = set([tuple(x) for x in lst2])\n",
    "    l = np.array([x for x in aset - bset])\n",
    "    return l\n",
    "\n",
    "def is_nan(x):\n",
    "    return (x is np.nan or x != x)\n",
    "\n",
    "\n",
    "def get_colors(num_colors):\n",
    "    colors=[]\n",
    "    for i in np.arange(0., 360., 360. / num_colors):\n",
    "        r_c = np.random.rand()\n",
    "        hue = i/360.\n",
    "        lightness = (50 + r_c * 10)/100.\n",
    "        saturation = (90 + r_c * 10)/100.\n",
    "        colors.append(colorsys.hls_to_rgb(hue, lightness, saturation))\n",
    "    return colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Set the variables \n",
    "\n",
    "colours=[ \"darkviolet\", \"red\", \"orange\", \"limegreen\", \"blue\", \"purple\", \"seagreen\",\"gold\",\"lightpink\",\"thistle\",\"mistyrose\",\"saddlebrown\",\"slategrey\",\n",
    "            \"palevioletred\",\"mediumvioletred\",\"yellowgreen\",\"darkolivegreen\",\"lemonchiffon\",\"chocolate\",\"lightsalmon\",\"lightcyan\",\"lightblue\", \"black\"]\n",
    "\n",
    "# perspective projection of stage coord to image coord for MERFISH\n",
    "\n",
    "pts_src = np.array([[-3357.4,-4078.2], [-3357.4,-3855.3], [-3134.5,-4078.2], [-3134.5,-3855.3]])\n",
    "# corresponding points from image 2 (i.e. (154, 174) matches (212, 80))\n",
    "pts_dst = np.array([[0.0,0.0], [0.0,2048.0], [2048.0,0.0],[2048.0,2048.0]])\n",
    "\n",
    "\n",
    "# calculate matrix H\n",
    "h, status = cv2.findHomography(pts_src, pts_dst)\n",
    "\n",
    "\n",
    "\n",
    "#set clustering algorithm choice\n",
    "bayes_MM = 0 # to run the DPMM else set to 0 for phenograph\n",
    "\n",
    "num_comp = 10 \n",
    "\n",
    "neigh_size = 15 #number of neighbours\n",
    "\n",
    "dist_ub = 150 #number of neighbours lie within this upper bound\n",
    "\n",
    "#dist_threshold = 100\n",
    "\n",
    "dpi_set = 100\n",
    "\n",
    "num_pg_k = 50 #number of PG clusters\n",
    "\n",
    "num_fov = 1 # change to total number of FoVs \n",
    "\n",
    "#sc_norm = 'zscore' # or 'log'\n",
    "#random.seed(9001)\n",
    "\n",
    "mRNA_asgn_counter = np.zeros([num_fov,1])\n",
    "\n",
    "s_size = 20\n",
    "\n",
    "#num_colour = 50\n",
    "win_size = 150 #vary between 70 - 150 to construct 'mock' cell\n",
    "#cm = plt.cm.get_cmap('tab20',num_colour)\n",
    "#cm = get_colors(num_colour)\n",
    "cm = colours\n",
    "\n",
    "f_size=10\n",
    "\n",
    "fig2p = 0\n",
    "\n",
    "z = 2\n",
    "\n",
    "num_cell_fov = np.zeros(num_fov, dtype=int)\n",
    "\n",
    "asgn_stats_list=[] #added \n",
    "asgn_stats_listoflists = [] #added \n",
    "\n",
    "mRNA_assign_listoflists = [] # 1st July 2020\n",
    "iterations = 2\n",
    "num_cores = multiprocessing.cpu_count() - 1\n",
    "inputs = range(num_fov)\n",
    "\n",
    "Merfish_genes = 140\n",
    "\n",
    "givens = True\n",
    "wgt_mock_cell = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Merfish assignment code\n",
    "## Code author SP\n",
    "##\n",
    "##\n",
    "## 5th Feb 2019 - clustering methods\n",
    "## 6th Feb 2019 - cluster moments extraction\n",
    "## 9th Feb 2019 - normality tests check, find nearest cells to an mRNA\n",
    "## 10th Feb 2019 - adding Gaussian copula to work with marginals\n",
    "## 10th Mar 2019 - loop added to take in multiple FOVs\n",
    "## 20th Mar 2019 - tidying up plots\n",
    "## 11th July 2019 - parallelising the loops\n",
    "## does not use the scRNA seq cluster moments \n",
    "##\n",
    "##\n",
    "##June - July 2020 - redoing the code with edits, adding Merfish start, distance ub, ##\n",
    "## July 31st 2020 - only printing the assignments in FOVs in loop Overall\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "############\n",
    "#### Main code\n",
    "############\n",
    "t1 = time.time()\n",
    "\n",
    "path_temp = os.path.join(path_data + '/barcode_metadata.csv')\n",
    "mRNA_metadata_t = np.array(pd.read_csv(path_temp).as_matrix())\n",
    "\n",
    "Merfish_CM_list = []\n",
    "mRNA_assign_list = []\n",
    "closest_cells_list = []\n",
    "mRNA_coords_list = []\n",
    "CoM_image_coords_list =[]\n",
    "cell_metadata_t_list = []\n",
    "mRNA_all_list = []\n",
    "\n",
    "num_cell_fov = np.zeros(num_fov, dtype=int) \n",
    "\n",
    "###########################################\n",
    "######## Build the mRNA stats per FOV #####\n",
    "###########################################\n",
    "\n",
    "inputs = range(num_fov) \n",
    "\n",
    "def processFOV(f):\n",
    "    print('Fov is: ', str(f))\n",
    "    \n",
    "    if (f < 10):\n",
    "        f_name = str(\"/cell_metadata/cell_metadata_fov_00\"+str(f)+\".csv\")\n",
    "    elif (f <= 99):\n",
    "        f_name = str(\"/cell_metadata/cell_metadata_fov_0\"+str(f)+\".csv\")\n",
    "    else:\n",
    "        f_name = str(\"/cell_metadata/cell_metadata_fov_\"+str(f)+\".csv\")\n",
    "    \n",
    "    mRNA_all = np.where(mRNA_metadata_t[:,1] == f) \n",
    "    mRNA_all = np.array(mRNA_all).flatten()\n",
    "    print(\"number of mRNA rows in Fov \",str(f),\"= \", mRNA_all.shape)\n",
    "    #return [mRNA_all_list.append(mRNA_all)]\n",
    "    \n",
    "    #read in mRNA needed for clustering\n",
    "    #mRNA_certain = np.where((mRNA_metadata_t[:,1] == f) & (mRNA_metadata_t[:,10] ==1))# & (mRNA_metadata_t[:,5] ==1.5)) #fov_id = 0 and in_feature = 1 and z-plane = 2\n",
    "    #mRNA_certain = np.array(mRNA_certain).flatten()\n",
    "    #print(\"number of mRNA rows in Fov \",str(f),\" for clustering = \", mRNA_certain.shape)\n",
    "\n",
    "    #print(\"number of unique genes or barcodes in Fov \",str(f),\"(for clustering) = \", np.unique(mRNA_metadata_t[mRNA_certain,0]).shape)\n",
    "    \n",
    "    \n",
    "    #read in mRNA needed for reassignment\n",
    "    mRNA_assign = np.where((mRNA_metadata_t[:,1] == f) & (mRNA_metadata_t[:,10] ==0))# & (mRNA_metadata_t[:,5] ==1.5)) #fov_id = 0 and in_feature = 0 and z-plane = 2\n",
    "    mRNA_assign = np.array(mRNA_assign).flatten()\n",
    "    print(\"number of mRNA rows in Fov \",str(f),\" for reassignment = \", mRNA_assign.shape)\n",
    "\n",
    "    print(\"number of unique genes or barcodes in Fov \",str(f),\" (for reassignment) = \", np.unique(mRNA_metadata_t[mRNA_assign,0]).shape)\n",
    "    #mRNA_assign_list.append(mRNA_assign)\n",
    "\n",
    "    # Build the count matrix of cells x genes per FOV\n",
    "    \n",
    "    path_temp = os.path.join(path_data + f_name)\n",
    "    \n",
    "    cell_metadata_t = np.array(pd.read_csv(path_temp).as_matrix()) #.astype(\"float\")\n",
    "    #cell_metadata_t_list.append(cell_metadata_t)\n",
    "    \n",
    "    num_cells = cell_metadata_t.shape[0]\n",
    "    num_genes = np.unique(mRNA_metadata_t[mRNA_all,0]).shape[0]\n",
    "\n",
    "    Merfish_CM = np.zeros([num_cells, num_genes], dtype=float) #this has to be a matrix with ALL the genes \n",
    "    \n",
    "    \n",
    "    for i in range(num_cells):\n",
    "        feat_id = cell_metadata_t[i,1]\n",
    "        #print(\"cell is\", feat_id)\n",
    "        CM_row = np.where(cell_metadata_t[:,1]==feat_id)\n",
    "        #print(\"cell row is\", CM_row)\n",
    "        #for 'f' fov, all the spots that are to be considered for clustering and for a given cell (feat_id)\n",
    "        feat_rows = np.where((mRNA_metadata_t[:,1]==f) & (mRNA_metadata_t[:,9]==feat_id) & (mRNA_metadata_t[:,10]==1))# & (mRNA_metadata_t[:,5] == 1.5)) #fov id, which cell and in/out of cell\n",
    "        #feat_rows = np.where((mRNA_metadata_t[:, 1] == 0) & (mRNA_metadata_t[:, 9] == feat_id))\n",
    "        feat_rows = np.array(feat_rows).flatten()\n",
    "        num_rows = feat_rows.shape[0]\n",
    "\n",
    "\n",
    "        for j in range(num_rows):\n",
    "            bc_id = mRNA_metadata_t[feat_rows[j],0]-1 #telling which gene it is and where the entry should go to in the count matrix\n",
    "            #print(\"gene is\", bc_id)\n",
    "            #Merfish_CM[CM_row,int(bc_id)] = Merfish_CM[CM_row,int(bc_id)]  +  mRNA_metadata_t[feat_rows[j],2]/mRNA_metadata_t[feat_rows[j],6] #area normalised values\n",
    "\n",
    "            #intensity-based count matrix\n",
    "            #Merfish_CM[CM_row,int(bc_id)] = Merfish_CM[CM_row,int(bc_id)]  +  mRNA_metadata_t[feat_rows[j],2]\n",
    "\n",
    "            #count matrix\n",
    "            Merfish_CM[CM_row,int(bc_id)] = Merfish_CM[CM_row,int(bc_id)]  +  1\n",
    "\n",
    "    #Merfish_CM_list.append(Merfish_CM)\n",
    "    \n",
    "    return f_name, mRNA_all, mRNA_assign, cell_metadata_t, num_cells, num_genes, Merfish_CM\n",
    "#num_cores = multiprocessing.cpu_count()\n",
    "     \n",
    "t1 = time.time()\n",
    "\n",
    "results = Parallel(n_jobs=num_cores)(delayed(processFOV)(f) for f in inputs)\n",
    "\n",
    "t2 = time.time()\n",
    "print('Time taken to execute FOV consolidation:', str(t2-t1),'secs')\n",
    "print('Time taken to execute FOV consolidation:', str((t2-t1)/60),'mins')\n",
    "\n",
    "#collect the results\n",
    "\n",
    "ff_name=[item[0] for item in results]\n",
    "mRNA_all_list=[item[1] for item in results]\n",
    "mRNA_assign_list=[item[2] for item in results]\n",
    "cell_metadata_t_list=[item[3] for item in results]\n",
    "num_cells_list=[item[4] for item in results]\n",
    "num_genes_list=[item[5] for item in results]\n",
    "Merfish_CM_list=[item[6] for item in results]\n",
    "\n",
    "## added 20th June 2020\n",
    "total_mRNA = 0\n",
    "total_dang_mRNA = 0\n",
    "for kl in range(num_fov):\n",
    "    total_mRNA = total_mRNA + mRNA_all_list[kl].shape[0]\n",
    "    total_dang_mRNA = total_dang_mRNA + mRNA_assign_list[kl].shape[0]\n",
    "print('total mRNA across all FOVs:', str(total_mRNA))\n",
    "print('total dangling mRNA across all FOVs:',str(total_dang_mRNA))\n",
    "###\n",
    "\n",
    "cell_cumsum = np.cumsum(num_cells_list)\n",
    "cell_cumsum = np.hstack((0,cell_cumsum))\n",
    "\n",
    "#1st July 2020\n",
    "Merfish_CM = np.zeros((1,Merfish_genes)) #will have all the Count matrices consolidated\n",
    "#Merfish_CM = np.zeros((1,num_genes_list[0])) #will have all the Count matrices consolidated\n",
    "\n",
    "for f in range(num_fov):\n",
    "    print(\"Number of cells in Fov \"+str(f)+\" is:\"+str(Merfish_CM_list[f].shape[0]))\n",
    "    Merfish_CM = np.concatenate((Merfish_CM,Merfish_CM_list[f]),axis=0)\n",
    "   \n",
    "    \n",
    "Merfish_CM = np.delete(Merfish_CM, (0), axis=0)\n",
    "\n",
    "#remove rows that have 0 rowsum\n",
    "\n",
    "row_sums_CM = Merfish_CM.sum(axis=1,keepdims=True)\n",
    "rows_remove = np.where(row_sums_CM==0)\n",
    "Merfish_CM = np.delete(Merfish_CM,(rows_remove),axis=0)\n",
    "\n",
    "#### added 20th June 2020\n",
    "print(\"rows_remove\")\n",
    "print(rows_remove)\n",
    "print('cell_cumsum b4')\n",
    "print(cell_cumsum)\n",
    "#adjust for cells removed per bin so that cumsum reflects the correct number of cells per FOV\n",
    "for i in range(len(rows_remove[0])):\n",
    "    print('i is', str(i))\n",
    "    for j in range(i+1):\n",
    "        print('j is', str(j))\n",
    "        if ((j+1) < (len(cell_cumsum))):\n",
    "            if (rows_remove[0][i] < cell_cumsum[j+1]):\n",
    "                cell_cumsum[j+1] = cell_cumsum[j+1] - 1\n",
    "                print(cell_cumsum)\n",
    "        continue\n",
    "    continue\n",
    "\n",
    "print('cell_cumsum after')\n",
    "print(cell_cumsum)\n",
    "##################\n",
    "M = zscore(Merfish_CM,axis=1)\n",
    "\n",
    "M_upd = M\n",
    "\n",
    "num_cells = M.shape[0]\n",
    "num_genes = Merfish_genes #M.shape[1]\n",
    "\n",
    "print(M.shape) # added 20th June 2020\n",
    "\n",
    "df_data = pd.DataFrame(Merfish_CM)\n",
    "df_data.to_csv(os.path.join(path_python + '/count_data/count_matrix_JeffMerfish_fovall.csv'), header=None, index=None)\n",
    "\n",
    "\n",
    "df_data = pd.DataFrame(M)\n",
    "df_data.to_csv(os.path.join(path_python + '/count_data/count_matrix_M_fovall.csv'), header=None, index=None)\n",
    "\n",
    "\n",
    "tsne = TSNE(n_components=2, verbose=1)#, perplexity=30, n_iter=900)\n",
    "X_2d = tsne.fit_transform(M)\n",
    "\n",
    "if(bayes_MM):\n",
    "    Mod_dpgmm=BayesianGaussianMixture(n_components=num_comp, covariance_type='full',weight_concentration_prior_type='dirichlet_process').fit(M)\n",
    "    cluster_asgn = Mod_dpgmm.predict(M)\n",
    "else:\n",
    "    communities, graph, Q = pg.cluster(M)#, k=num_comp, min_cluster_size=1)\n",
    "    cluster_asgn = communities\n",
    "\n",
    "df_clusterlabel = pd.DataFrame(cluster_asgn)\n",
    "df_clusterlabel.to_csv(os.path.join(path_python + '/count_data/cluster_labels_b4.csv'), header=None,\n",
    "                   index=None)\n",
    "\n",
    "unique, counts = np.unique(cluster_asgn, return_counts=True)\n",
    "\n",
    "#t-SNE plots\n",
    "target_ids = range(len(unique))\n",
    "plt.figure(figsize=(8,8),frameon=False,dpi=dpi_set)\n",
    "plt.axis('off')\n",
    "#colors = 'r', 'g', 'b', 'c', 'm', 'y', 'k', 'w', 'orange', 'purple'\n",
    "for i,  label in zip(target_ids,  unique):\n",
    "    plt.scatter(X_2d[ cluster_asgn==(i), 0], X_2d[ cluster_asgn==(i), 1],  color=cm[i], label=label,s=s_size)\n",
    "plt.legend()\n",
    "plt.title('t-SNE')\n",
    "plt.show()\n",
    "file_name = os.path.join(path_python+'/figures_python/tSNE_labels.png')\n",
    "plt.savefig(file_name,dpi=dpi_set)\n",
    "\n",
    "df_data = pd.DataFrame(X_2d)\n",
    "df_data.to_csv(os.path.join(path_python + '/count_data/tSNE_embedding_pre.csv'), header=None, index=None)\n",
    "\n",
    "###############################\n",
    "## construct the cluster moments\n",
    "###############################\n",
    "\n",
    "mean_mat = [] #1 x genes per entry. K x genes overall\n",
    "cov_list = [] # genes x genes per entry. K genes x genes overall\n",
    "joint_mat = [] # samples X genes per entry. K samples x genes overall\n",
    "U_mat = [] # samples X genes per entry. K samples x genes overall\n",
    "num_genes = Merfish_genes ## \n",
    "\n",
    "\n",
    "for i in range(len(unique)): #target_ids:\n",
    "    print(i)\n",
    "    rows = np.where(cluster_asgn==unique[i])\n",
    "    rows = np.array(rows).flatten()\n",
    "\n",
    "\n",
    "    if (rows.shape[0] > 1):\n",
    "        mean_mat.append(np.mean(M[rows, :], axis=0)) # added 3rd july \n",
    "        #cov_list.append(np.cov(M[rows, 0:num_genes].T))\n",
    "        #joint_mat.append(np.random.multivariate_normal(mean_mat[i][0:num_genes], cov_list[i], num_unif_samples))\n",
    "        #####joint_mat.append(multivariate_normal.logpdf(mean_mat[i][0:10], cov_list[i], num_unif_samples))\n",
    "        cov_mat = np.cov(M[rows, 0:num_genes].T)\n",
    "        cov_mat = np.dot(cov_mat,cov_mat.T)\n",
    "        np.fill_diagonal(cov_mat, np.diag(cov_mat)+0.01)\n",
    "        if (givens):\n",
    "            (Q, R) = givens_rotation(cov_mat)\n",
    "            cov_list.append(R)  \n",
    "        else:\n",
    "            cov_list.append(cov_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mRNA_assign_list[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################################\n",
    "## Iteration 0 #################\n",
    "################################\n",
    "\n",
    "\n",
    "#inputs = range(num_fov) \n",
    "def processIter0(f):\n",
    "    \n",
    "    #mRNA_asgn_counter = np.zeros([num_fov,1]) #added 1st July 2020\n",
    "    asgn_stats = np.array([0,0,0,0]) #added , iter, fov, p, cellitwentto\n",
    "\n",
    "    print('Fov is: ', str(f))\n",
    "\n",
    "    num_cells = cell_metadata_t_list[f].shape[0]\n",
    "\n",
    "    #perspective projection\n",
    "    if z==1:\n",
    "        z_x = 6\n",
    "        z_y = 7\n",
    "    else:\n",
    "        z_x = 6 + 2**(z-1)\n",
    "        z_y = 7 + 2**(z-1)\n",
    "\n",
    "    rw_x_boundary = cell_metadata_t_list[f][:,z_x]\n",
    "    rw_y_boundary = cell_metadata_t_list[f][:,z_y]\n",
    "\n",
    "    CoM_stage_coords = np.zeros([num_cells, 2], dtype=np.float)\n",
    "    CoM_image_coords = np.zeros([num_cells, 2], dtype=np.float)\n",
    "\n",
    "    #fig1 = (f+1)*100+(f+1)\n",
    "    \n",
    "    '''nx31\n",
    "    if(fig2p==1):\n",
    "        fig2 = (f+1)*200+(f+1)\n",
    "    fig4 = (f+1)*400+(f+1)\n",
    "\n",
    "    plt.figure(fig4,dpi=dpi_set,figsize=(12,12))\n",
    "    plt.title(str(f))\n",
    "    plt.axis('off')\n",
    "    '''\n",
    "    \n",
    "    print('plotting cells')\n",
    "    for i in range(num_cells): #i will be i+2th row in the xls\n",
    "\n",
    "        #print(\"Cell= \" + str(i))\n",
    "\n",
    "        if(is_nan(rw_x_boundary[i]) | is_nan(rw_y_boundary[i])): #entries were empty: no segmentation found\n",
    "            print(\"skip cell \", str(i))\n",
    "        else:\n",
    "            m = rw_x_boundary[i].split(';')\n",
    "            n = rw_y_boundary[i].split(';')\n",
    "            num_coord = m.__len__()\n",
    "\n",
    "            # this can be done smarter. for now I cast every string float to float and place into a list\n",
    "            m_list = []\n",
    "            n_list = []\n",
    "            for j in range(num_coord-1):\n",
    "                if( (float(m[j])!=float(m[j]))  or (float(n[j])!=float(n[j]))):\n",
    "                    print(\"Coord NaN is at = \" + str(j))\n",
    "                    #continue\n",
    "                else:\n",
    "                    m_list.append(float(m[j]))\n",
    "                    n_list.append(float(n[j]))\n",
    "\n",
    "            #to plot the stage coords\n",
    "            #for k in np.arange(0, m_list.__len__()-1, 2):\n",
    "                #print(k)\n",
    "                #plt.figure(fig1,figsize=(10,10),dpi=dpi_set)\n",
    "            #    connectpoints(m_list, n_list, k, k + 1)\n",
    "\n",
    "            x_coord = np.array(m_list)  # (256,)\n",
    "            y_coord = np.array(n_list)  # (256,)\n",
    "            f1 = np.vstack((x_coord, y_coord))  # (2, 256)\n",
    "\n",
    "            cX = sum(f1[0, :]) / f1.shape[1]\n",
    "            cY = sum(f1[1, :]) / f1.shape[1]\n",
    "\n",
    "            '''nx31\n",
    "            CoM_stage_coords[i, 0] = cX\n",
    "            CoM_stage_coords[i, 1] = cY\n",
    "            '''\n",
    "            \n",
    "            #plt.figure(fig1,dpi=dpi_set,figsize=(10,10))\n",
    "            #axes = plt.gca()\n",
    "            #axes.add_patch(Polygon(np.transpose(f1),closed=True, facecolor=colours[cluster_asgn[i]]))\n",
    "            #plt.axis('off')\n",
    "\n",
    "\n",
    "            #to project to the image coords\n",
    "\n",
    "            im_list = []\n",
    "            in_list = []\n",
    "            for s in range(m_list.__len__()):\n",
    "                a = np.array([[f1[0, s], f1[1, s]]], dtype='float32')\n",
    "                a = np.array([a])\n",
    "                pointsOut = cv2.perspectiveTransform(a, h)\n",
    "                im_list.append(pointsOut[0,0,0])\n",
    "                in_list.append(pointsOut[0,0,1])\n",
    "\n",
    "            '''\n",
    "            for k in np.arange(0, im_list.__len__()-1, 2):\n",
    "\n",
    "                plt.figure(fig2,figsize=(10,10),dpi=dpi_set)\n",
    "                connectpoints(im_list, in_list, k, k + 1)\n",
    "            '''\n",
    "            ix_coord = np.array(im_list)  # (256,)\n",
    "            iy_coord = np.array(in_list)  # (256,)\n",
    "            f_image = np.vstack((ix_coord, iy_coord))  # (2, 256)\n",
    "\n",
    "            cX_image = sum(f_image[0, :]) / f_image.shape[1]\n",
    "            cY_image = sum(f_image[1, :]) / f_image.shape[1]\n",
    "\n",
    "            CoM_image_coords[i,0] = cX_image\n",
    "            CoM_image_coords[i,1] = cY_image\n",
    "\n",
    "\n",
    "            if(fig2p==1):\n",
    "                plt.figure(fig2) #,dpi=dpi_set,figsize=(10,10))\n",
    "                axes = plt.gca()\n",
    "                #axes.add_patch(Polygon(np.transpose(f_image),closed=True, facecolor=colours[cluster_asgn[i]]))\n",
    "                axes.add_patch(Polygon(np.transpose(f_image),closed=True, facecolor=cm[cluster_asgn[i]]))\n",
    "                #plt.title('fig2')\n",
    "                #plt.axis('off')\n",
    "\n",
    "                 \n",
    "\n",
    "        '''nx31\n",
    "        ## plot mRNA inside the cell\n",
    "        ##\n",
    "        feat_id = cell_metadata_t_list[f][i,1]\n",
    "        feat_rows = np.where((mRNA_metadata_t[:,1]==f) & (mRNA_metadata_t[:,9]==feat_id) & (mRNA_metadata_t[:,10]==1)) #fov id, which cell and in/out of cell\n",
    "        #feat_rows = np.where((mRNA_metadata_t[:, 1] == 0) & (mRNA_metadata_t[:, 9] == feat_id))\n",
    "        feat_rows = np.array(feat_rows).flatten()\n",
    "        num_rows = feat_rows.shape[0]\n",
    "        for j in range(num_rows):\n",
    "            x_stage_coord = mRNA_metadata_t[feat_rows[j],3]\n",
    "            y_stage_coord = mRNA_metadata_t[feat_rows[j],4]\n",
    "            a = np.array([[x_stage_coord, y_stage_coord]], dtype='float32')\n",
    "            a = np.array([a])\n",
    "\n",
    "            #perspective proj\n",
    "            pointsOut = cv2.perspectiveTransform(a, h)\n",
    "\n",
    "            x_image_coord = pointsOut[0,0,0]\n",
    "            y_image_coord = pointsOut[0,0,1]\n",
    "\n",
    "            #plt.plot(x_image_coord, y_image_coord,'o',markersize=2, c=colours[cluster_asgn[i]],alpha=0.5)\n",
    "            if ((i+cell_cumsum[f]) < M.shape[0]):#added 20th June 2020            \n",
    "                plt.plot(x_image_coord, y_image_coord,'.',markersize=4, color=cm[cluster_asgn[i+cell_cumsum[f]]])   \n",
    "        '''\n",
    "\n",
    "    #CoM_image_coords_list.append(CoM_image_coords)\n",
    "\n",
    "    # find all mRNA coords after projection\n",
    "    print(\"find all mRNA coords after projection\")\n",
    "    l2_norm_mat = np.zeros([num_cells, len(mRNA_assign_list[f])],dtype=float)\n",
    "    closest_cells = []\n",
    "    #getting all the mRNA coords in that specific fov\n",
    "    mRNA_coords = np.zeros([len(mRNA_all_list[f]), 2], dtype=np.float)\n",
    "\n",
    "    for j in range(len(mRNA_all_list[f])):\n",
    "\n",
    "        #get mRNA CoM\n",
    "        m_x = mRNA_metadata_t[mRNA_all_list[f][j],3]\n",
    "        m_y = mRNA_metadata_t[mRNA_all_list[f][j],4]\n",
    "        a = np.array([[m_x, m_y]], dtype='float32')\n",
    "        a = np.array([a])\n",
    "\n",
    "        #perspective proj\n",
    "        pointsOut = cv2.perspectiveTransform(a, h)\n",
    "\n",
    "        mRNA_coords[j, 0] = pointsOut[0,0,0]\n",
    "        mRNA_coords[j, 1] = pointsOut[0,0,1]\n",
    "\n",
    "    #mRNA_coords_list.append(mRNA_coords)  #added \n",
    "\n",
    "    # all mRNA to be assigned\n",
    "    asgn_mRNA = np.intersect1d(mRNA_assign_list[f],mRNA_all_list[f])\n",
    "\n",
    "    # per mRNA to be assigned, do:\n",
    "    print(\"per mRNA to be assigned do: loop\")\n",
    "    for l in range(0,len(asgn_mRNA)):\n",
    "\n",
    "        # mock cell creation\n",
    "        # print(\"create mock cell\")\n",
    "        #print(l)\n",
    "        p = np.array(np.where(asgn_mRNA[l]==mRNA_all_list[f])).flatten()\n",
    "        xc = mRNA_coords[p, 0]   \n",
    "        yc = mRNA_coords[p, 1]\n",
    "        l_xc = xc - win_size\n",
    "        r_xc = xc + win_size\n",
    "        l_yc = yc - win_size\n",
    "        u_yc = yc + win_size\n",
    "\n",
    "        #ys = mRNA_coords[:,1][(mRNA_coords[:,1] >= l_yc) & (mRNA_coords[:,1] <= u_yc)]\n",
    "        #xs = [np.where((mRNA_coords[:,0] >= l_xc) & (mRNA_coords[:,0] <= r_xc))]\n",
    "        #ys = [np.where((mRNA_coords[:,1] >= l_yc) & (mRNA_coords[:,1] <= u_yc))]\n",
    "\n",
    "        # finding its nearest cells\n",
    "        #print(\"find its nearest cells\")\n",
    "        for i in range(num_cells):\n",
    "            l2_norm_mat[i,l] = distance.euclidean([xc,yc], [CoM_image_coords[i,0],CoM_image_coords[i,1]])\n",
    "        ##21st June 2020 check \n",
    "        #print('distances 1')\n",
    "        #print(l2_norm_mat[0:10,0:10])\n",
    "        #23rd June 2020\n",
    "        #closest_cells.append(np.argsort(l2_norm_mat[:,l])[0:neigh_size]) #this is the order of cells (ie cell id) but in Python. For real world, +1\n",
    "        closest_cells.append(np.argsort(l2_norm_mat[:,l])[np.where((np.sort(l2_norm_mat[:,l])[0:neigh_size]<dist_ub)==True)])\n",
    "        #1st quadrant\n",
    "        ##print(\"build the mock cell\")\n",
    "        #for the circle\n",
    "        ind = np.where(pow(mRNA_coords[:,0]-xc,2) + pow(mRNA_coords[:,1]-yc,2) <= pow(win_size,2))\n",
    "        #ind = [np.where((mRNA_coords[:,1] >= l_yc) & (mRNA_coords[:,1] <= u_yc) & (mRNA_coords[:,0] >= l_xc) & (mRNA_coords[:,0] <= r_xc))]\n",
    "        ind = np.array(ind).flatten()\n",
    "        num_mRNA_neighs = ind.shape[0]\n",
    "\n",
    "        ##########mockcell weighted\n",
    "        \n",
    "        l2_norm_mat_1 =np.zeros([1, num_mRNA_neighs],dtype=float)\n",
    "        for i in range(num_mRNA_neighs):\n",
    "            l2_norm_mat_1[0,i] = distance.euclidean([xc,yc], [mRNA_coords[ind[i],0],mRNA_coords[ind[i],1]])\n",
    "\n",
    "        \n",
    "        mock_cell = np.zeros((1,num_genes))  #note change\n",
    "        \n",
    "        #####weighted_mc - 9th sep 2020\n",
    "        if (wgt_mock_cell):\n",
    "            for k in range(num_mRNA_neighs):\n",
    "                bc_id = mRNA_metadata_t[mRNA_all_list[f][ind[k]],0]-1 #telling which gene it is and where the entry should go to in the count matrix\n",
    "                if(bc_id < num_genes):\n",
    "                    mock_cell[0,int(bc_id)] = mock_cell[0,int(bc_id)]  +  1/l2_norm_mat_1[0,k] # 9th sep 2020\n",
    "\n",
    "        \n",
    "        else: #unweighted mock cell\n",
    "            for k in range(num_mRNA_neighs):\n",
    "                bc_id = mRNA_metadata_t[mRNA_all_list[f][ind[k]],0]-1 #telling which gene it is and where the entry should go to in the count matrix\n",
    "                if(bc_id < num_genes):\n",
    "                    mock_cell[0,int(bc_id)] = mock_cell[0,int(bc_id)]  +  1 # 9th sep 2020\n",
    "\n",
    "        '''\n",
    "        #mock_cell = np.zeros((1,num_genes))  \n",
    "        #num_genes = 10\n",
    "        mock_cell = np.zeros((1,num_genes))  #note change\n",
    "\n",
    "        for k in range(num_mRNA_neighs):\n",
    "            bc_id = mRNA_metadata_t[mRNA_all_list[f][ind[k]],0]-1 #telling which gene it is and where the entry should go to in the count matrix\n",
    "            if(bc_id < num_genes):\n",
    "                mock_cell[0,int(bc_id)] = mock_cell[0,int(bc_id)]  +  1\n",
    "\n",
    "        #mock_cell = mock_cell[:,0:10] #note change //\n",
    "\n",
    "        '''\n",
    "        ########\n",
    "        \n",
    "\n",
    "        # MLE with all K clusters\n",
    "        #print(\"computing ML\")\n",
    "        mle_mockcell = np.zeros((1,len(unique)))\n",
    "\n",
    "        for u in range(len(unique)):\n",
    "            #mle_mockcell[0,u] = multivariate_normal.logpdf(zscore(mock_cell[0,0:10],axis=0), mean_mat[u], cov_list[u])\n",
    "            mle_mockcell[0,u] = multivariate_normal.logpdf(zscore(mock_cell,axis=1), mean_mat[u][0:num_genes], cov_list[u])\n",
    "        #print(mle_mockcell)\n",
    "\n",
    "        mle_k = np.argmin(mle_mockcell)       \n",
    "\n",
    "        #assign to closest cell of that ML cluster\n",
    "\n",
    "        cc = np.array(closest_cells[l]).flatten()\n",
    "        mle_cc = np.array(np.where(cluster_asgn[cc]==mle_k)).flatten()\n",
    "\n",
    "\n",
    "        #plot this line from mRNA to assign to cell \n",
    "        if (len(mle_cc)!=0): #means there are cells in the neighbourhood of mRNA that are of the same class\n",
    "            cell_to_map = cc[mle_cc[0]]\n",
    "            #np.where(cell_metadata_t_list[1][cell_to_map,1] ==cell_metadata_t_list[1][:,1]) = cell_to_map\n",
    "            # meaning cell_to_map indexes cell_metadata_t[f]\n",
    "            #plt.figure(fig4,dpi=dpi_set,figsize=(12,12))\n",
    "            if ((cell_to_map+cell_cumsum[f]) < M.shape[0]): #added 20th June 2020\n",
    "                '''nx31\n",
    "                plt.plot(xc, yc,'x',markersize=4, c=cm[cluster_asgn[cell_to_map+cell_cumsum[f]]],alpha=0.5)\n",
    "\n",
    "                x_temp = [CoM_image_coords[cell_to_map,0],xc]\n",
    "                y_temp = [CoM_image_coords[cell_to_map,1],yc]\n",
    "                '''\n",
    "                mRNA_asgn_counter[f] = mRNA_asgn_counter[f] + 1\n",
    "\n",
    "                #nx31 plt.plot(x_temp, y_temp, linewidth=1, c=cm[cluster_asgn[cell_to_map+cell_cumsum[f]]])\n",
    "\n",
    "            #if (f==0):\n",
    "            #    M_upd[cell_to_map, int(mRNA_metadata_t[p, 0])-1] = M_upd[cell_to_map, int(mRNA_metadata_t[p, 0])-1] + 1\n",
    "            #else:\n",
    "            #    M_upd[cell_to_map+int(Merfish_CM_list[f-1].shape[0]), int(mRNA_metadata_t[mRNA_all_list[f-1].shape+p, 0])-1] = M_upd[cell_to_map+int(Merfish_CM_list[f-1].shape[0]), int(mRNA_metadata_t[mRNA_all_list[f-1].shape+p, 0])-1] + 1\n",
    "\n",
    "            #M_upd[cell_metadata_t_list[f][cell_to_map,1]-1, int(mRNA_metadata_t[mRNA_all_list[f][p],0]-1)] = M_upd[cell_metadata_t_list[f][cell_to_map,1]-1, int(mRNA_metadata_t[mRNA_all_list[f][p],0]-1)] + 1\n",
    "            # -ve 1 only because I need to update a Python structure!\n",
    "                Merfish_CM_list[f][cell_to_map, int(mRNA_metadata_t[mRNA_all_list[f][p],0]-1)] = Merfish_CM_list[f][cell_to_map, int(mRNA_metadata_t[mRNA_all_list[f][p],0]-1)] + 1\n",
    "\n",
    "                a = np.array([0,f,p,cell_to_map]) #added \n",
    "                asgn_stats = np.vstack((asgn_stats,a)) #added \n",
    "        '''nx31\n",
    "        else: \n",
    "            #print(l)\n",
    "            #print('no match with any cluster')\n",
    "            #plt.figure(fig4)#,dpi=dpi_set,figsize=(12,12))\n",
    "            plt.plot(xc,yc, '.', markersize=4, c='black')#, alpha=0.5)\n",
    "        '''\n",
    "\n",
    "    '''nx31\n",
    "    plt.figure(fig4)\n",
    "    plt.title(\"Iter_\"+str(0)+\"_\"+\"Fov_\"+str(f))\n",
    "    plt.axis('off')\n",
    "    file_name = os.path.join(path_python+'/figures_python/mRNA_assigned_seg_all_Iter=0_Fov='+str(f)+'.png')\n",
    "    plt.savefig(file_name,dpi=dpi_set)\n",
    "    '''\n",
    "    asgn_stats = np.delete(asgn_stats, (0), axis=0) #added \n",
    "    #asgn_stats_list.append(asgn_stats) # added \n",
    "    \n",
    "    return  CoM_image_coords, mRNA_coords, asgn_stats,mRNA_asgn_counter #20th June 2020\n",
    "\n",
    "t3 = time.time()\n",
    "results1 = Parallel(n_jobs=num_cores)(delayed(processIter0)(f) for f in inputs)\n",
    "\n",
    "t4 = time.time()\n",
    "print('Time taken to perform Iter 0:', str(t4-t3),'secs')\n",
    "print('Time taken to perform Iter 0:', str((t4-t3)/60),'mins')\n",
    "\n",
    "CoM_image_coords_list=[item[0] for item in results1]\n",
    "mRNA_coords_list=[item[1] for item in results1]\n",
    "asgn_stats_list=[item[2] for item in results1]\n",
    "mRNA_asgn_counter_list=[item[3] for item in results1] #20th June 2020\n",
    "\n",
    "asgn_stats_listoflists.append(asgn_stats_list)\n",
    "\n",
    "#### added below from scrna-seq code on 20th June 2020\n",
    "# added \n",
    "# 17th, 18th Nov 2019\n",
    "\n",
    "sd = []\n",
    "for si in range(num_fov):\n",
    "    print(mRNA_asgn_counter_list[si])\n",
    "    if (np.all(mRNA_asgn_counter_list[si]==0)):\n",
    "        print(si)\n",
    "        sd.append(si)\n",
    "ss = np.delete(range(num_fov),sd)\n",
    "\n",
    "#s=np.array(mRNA_asgn_counter_list).flatten()\n",
    "#print('s is ', s)\n",
    "\n",
    "#ss = np.array(np.where(s!=0)).flatten()\n",
    "#s = s[ss]\n",
    "#mRNA_asgn_counter_list = s\n",
    "print('ss is',ss)\n",
    "print('len of ss ', len(ss))\n",
    "\n",
    "num_f = len(ss)\n",
    "print('num_f ', num_f)\n",
    "\n",
    "\n",
    "print('Total time to execute code across', str(iterations),'Iterations and',str(num_fov), 'FOVs is:', str((t4-t1)/60),'minutes')\n",
    "print('assigned stats in Iter 0 across FOVs are:', str(mRNA_asgn_counter_list))\n",
    "\n",
    "#print('Total time to execute code across', str(iterations),'Iterations and',str(num_fov), 'FOVs is:', str((t4-t1)/60),'minutes') \n",
    "\n",
    "#print('assigned stats in Iter 0 across FOVs are:', str(mRNA_asgn_counter_list))\n",
    "#print('number of mRNA to be assigned per FOV:', str(mRNA_assign_list))\n",
    "\n",
    "#num_f = num_fov - 1\n",
    "\n",
    "num_fov = len(ss)\n",
    "print('num_fov ', num_fov)\n",
    "\n",
    "for kk in range(0,num_f):\n",
    "    print('% of mRNA assigned over total dangling mRNAs in FOV ', str(ss[kk]),'is ',\n",
    "          str(int(len(mRNA_asgn_counter_list[ss[kk]]))/len(mRNA_assign_list[ss[kk]])))\n",
    "    print('% of mRNA assigned over total mRNAs in FOV ', str(ss[kk]),'is ', \n",
    "          str(int(len(mRNA_asgn_counter_list[ss[kk]]))/len(mRNA_all_list[ss[kk]])))\n",
    "\n",
    "####  added end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################################\n",
    "######### Iterations #####################\n",
    "##########################################\n",
    "\n",
    "\n",
    "def processItern(f):\n",
    "\n",
    "    print('Fov is: ', str(f))\n",
    "    asgn_stats = np.array([iter1,f,0,0]) # iter, fov, p, cellitwentto\n",
    "    mRNA_asgn_counter = 0\n",
    "\n",
    "    num_cells = cell_metadata_t_list[f].shape[0]\n",
    "\n",
    "    '''nx31\n",
    "    if(fig2p==1):\n",
    "        fig2 = (f+1)*200+(f+1)\n",
    "    #fig4 = (iter1+f+1)*4+(iter1+f+1)\n",
    "    fig4 = (iter1+f+1)*4+f\n",
    "    # cells are always the same per fov. so this is ok, no need to keep track of the iteration\n",
    "    plt.figure(fig4,dpi=dpi_set,figsize=(12,12))\n",
    "    plt.title(str(f))\n",
    "    plt.axis('off')\n",
    "    '''\n",
    "    CoM_im_coords = CoM_image_coords_list[f] \n",
    "\n",
    "    '''\n",
    "    for i in range(num_cells):\n",
    "        feat_id = cell_metadata_t_list[f][i,1]\n",
    "        feat_rows = np.where((mRNA_metadata_t[:,1]==f) & (mRNA_metadata_t[:,9]==feat_id) & (mRNA_metadata_t[:,10]==1)) #fov id, which cell and in/out of cell#feat_rows = np.where((mRNA_metadata_t[:, 1] == 0) & (mRNA_metadata_t[:, 9] == feat_id))\n",
    "        feat_rows = np.array(feat_rows).flatten()\n",
    "        num_rows = feat_rows.shape[0]\n",
    "        for j in range(num_rows):\n",
    "            x_stage_coord = mRNA_metadata_t[feat_rows[j],3]\n",
    "            y_stage_coord = mRNA_metadata_t[feat_rows[j],4]\n",
    "            a = np.array([[x_stage_coord, y_stage_coord]], dtype='float32')\n",
    "            a = np.array([a])\n",
    "            pointsOut = cv2.perspectiveTransform(a, h)\n",
    "            x_image_coord = pointsOut[0,0,0]\n",
    "            y_image_coord = pointsOut[0,0,1]\n",
    "            plt.plot(x_image_coord, y_image_coord,'o',markersize=2, color=cm.colors[cluster_asgn[i]])\n",
    "    '''\n",
    "\n",
    "    #find l2-norm of each gene candidate to every cell\n",
    "    #do just for image coordinates, it should be the same cells in the stage coords as wel\n",
    "    l2_norm_mat =np.zeros([num_cells, len(mRNA_assign_list[f])],dtype=float)\n",
    "    closest_cells = []\n",
    "\n",
    "    mRNA_im_coords = mRNA_coords_list[f]  #added  # all mRNA are always the same per fov. so this is ok, no need to keep track of the iteration\n",
    "\n",
    "    #new mRNA assign list per fov #added \n",
    "    to_rem_mRNA = asgn_stats_listoflists[iter1-1][f]\n",
    "    len_rem_mRNA = len(to_rem_mRNA)\n",
    "    print('len_rem_mRNA ',str(len_rem_mRNA))\n",
    "    to_rem_mRNA_1=np.array([0])\n",
    "    for l in range(len_rem_mRNA):\n",
    "        pp = np.array(np.where(mRNA_all_list[f][to_rem_mRNA[l,2]]==mRNA_assign_list[f])).flatten()\n",
    "        to_rem_mRNA_1 = np.concatenate((to_rem_mRNA_1,pp))\n",
    "\n",
    "    to_rem_mRNA_1=np.delete(to_rem_mRNA_1,(0),axis=0)\n",
    "    mRNA_assign_list[f]=np.delete(mRNA_assign_list[f],to_rem_mRNA_1,axis=0).copy() #added ,copy() 1st July 2020\n",
    "    print('mRNA_assign_list[f] ', str(len(mRNA_assign_list[f]))) \n",
    "    \n",
    "    # all mRNA to be assigned\n",
    "    asgn_mRNA = np.intersect1d(mRNA_assign_list[f],mRNA_all_list[f])\n",
    "    print('asgn_mRNA ',str(len(asgn_mRNA)))\n",
    "\n",
    "    # per mRNA to be assigned, do:\n",
    "    print(\"per mRNA to be assigned do: loop\")\n",
    "    for l in range(0,len(asgn_mRNA)): #//added 1st July 2020\n",
    "\n",
    "        p = np.array(np.where(asgn_mRNA[l]==mRNA_all_list[f])).flatten()\n",
    "        xc = mRNA_im_coords[p, 0]\n",
    "        yc = mRNA_im_coords[p, 1]\n",
    "        l_xc = xc - win_size\n",
    "        r_xc = xc + win_size\n",
    "        l_yc = yc - win_size\n",
    "        u_yc = yc + win_size\n",
    "        for i in range(num_cells):\n",
    "            l2_norm_mat[i,l] = distance.euclidean([xc,yc], [CoM_im_coords[i,0],CoM_im_coords[i,1]])\n",
    "        ##21st June 2020 check\n",
    "        #print('distances 1')\n",
    "        #print(l2_norm_mat[0:10,0:10])\n",
    "\n",
    "        #23rd June 2020\n",
    "        #closest_cells.append(np.argsort(l2_norm_mat[:,l])[0:neigh_size]) #this is the order of cells (ie cell id) but in Python. For real world, +1\n",
    "        closest_cells.append(np.argsort(l2_norm_mat[:,l])[np.where((np.sort(l2_norm_mat[:,l])[0:neigh_size]<dist_ub)==True)]) \n",
    "        #1st quadrant\n",
    "        #print(\"build the mock cell\")\n",
    "        \n",
    "        # for a square\n",
    "        # ind = [np.where((mRNA_im_coords[:,1] >= l_yc) & (mRNA_im_coords[:,1] <= u_yc) & (mRNA_im_coords[:,0] >= l_xc) & (mRNA_im_coords[:,0] <= r_xc))]\n",
    "        \n",
    "        # for a circle\n",
    "        ind = np.where(pow(mRNA_im_coords[:,0]-xc,2) + pow(mRNA_im_coords[:,1]-yc,2) <= pow(win_size,2))\n",
    "        ind = np.array(ind).flatten()\n",
    "        num_mRNA_neighs = ind.shape[0]\n",
    "\n",
    "        ##########mockcell weighted\n",
    "        \n",
    "        l2_norm_mat_1 =np.zeros([1, num_mRNA_neighs],dtype=float)\n",
    "        for i in range(num_mRNA_neighs):\n",
    "            l2_norm_mat_1[0,i] = distance.euclidean([xc,yc], [mRNA_im_coords[ind[i],0],mRNA_im_coords[ind[i],1]])\n",
    "\n",
    "        \n",
    "        mock_cell = np.zeros((1,num_genes))  #note change\n",
    "        \n",
    "        #####weighted_mc - 9th sep 2020\n",
    "        if (wgt_mock_cell):\n",
    "            for k in range(num_mRNA_neighs):\n",
    "                bc_id = mRNA_metadata_t[mRNA_all_list[f][ind[k]],0]-1 #telling which gene it is and where the entry should go to in the count matrix\n",
    "                if(bc_id < num_genes):\n",
    "                    mock_cell[0,int(bc_id)] = mock_cell[0,int(bc_id)]  +  1/l2_norm_mat_1[0,k] # 9th sep 2020\n",
    "\n",
    "        \n",
    "        else: #unweighted mock cell\n",
    "            for k in range(num_mRNA_neighs):\n",
    "                bc_id = mRNA_metadata_t[mRNA_all_list[f][ind[k]],0]-1 #telling which gene it is and where the entry should go to in the count matrix\n",
    "                if(bc_id < num_genes):\n",
    "                    mock_cell[0,int(bc_id)] = mock_cell[0,int(bc_id)]  +  1 # 9th sep 2020\n",
    "\n",
    "        '''\n",
    "        #mock_cell = np.zeros((1,num_genes))  \n",
    "        #num_genes = 10\n",
    "        mock_cell = np.zeros((1,num_genes))\n",
    "\n",
    "        for k in range(num_mRNA_neighs):\n",
    "\n",
    "            bc_id = mRNA_metadata_t[mRNA_all_list[f][ind[k]],0]-1 #telling which gene it is and where the entry should go to in the count matrix\n",
    "            if(bc_id < num_genes):\n",
    "                mock_cell[0,int(bc_id)] = mock_cell[0,int(bc_id)]  +  1\n",
    "\n",
    "        '''\n",
    "        ########\n",
    "        \n",
    "\n",
    "        # MLE with all K clusters\n",
    "        #print(\"computing ML\")\n",
    "        mle_mockcell = np.zeros((1,len(unique)))\n",
    "\n",
    "        #for u in range(len(unique)):\n",
    "        for u in range(len(mean_mat)): \n",
    "            #mle_mockcell[0,u] = multivariate_normal.logpdf(zscore(mock_cell[0,0:10],axis=0), mean_mat[u], cov_list[u])\n",
    "            mle_mockcell[0,u] = multivariate_normal.logpdf(zscore(mock_cell,axis=1), mean_mat[u][0:num_genes], cov_list[u])#print(mle_mockcell)\n",
    "\n",
    "        mle_k = np.argmin(mle_mockcell)\n",
    "        cc = np.array(closest_cells[l]).flatten()\n",
    "        mle_cc = np.array(np.where(cluster_asgn[cc]==mle_k)).flatten()\n",
    "\n",
    "        #plot this line from mRNA to assign to cell \n",
    "        if (len(mle_cc)!=0): #means there are cells in the neighbourhood of mRNA that are of the same class\n",
    "            cell_to_map = cc[mle_cc[0]]\n",
    "            #plt.figure(fig4,dpi=dpi_set,figsize=(12,12))\n",
    "            if ((cell_to_map+cell_cumsum[f]) < M_upd.shape[0]): #added 20th June 2020\n",
    "                '''nx31\n",
    "                plt.plot(xc, yc,'x',markersize=4, c=cm[cluster_asgn[cell_to_map+cell_cumsum[f]]],alpha=0.5)\n",
    "\n",
    "                x_temp = [CoM_im_coords[cell_to_map,0],xc]\n",
    "                y_temp = [CoM_im_coords[cell_to_map,1],yc]\n",
    "                '''\n",
    "                mRNA_asgn_counter = mRNA_asgn_counter + 1\n",
    "\n",
    "                #nx31 plt.plot(x_temp, y_temp, linewidth=1, c=cm[cluster_asgn[cell_to_map+cell_cumsum[f]]])\n",
    "                #Merfish_CM[cell_to_map+cell_cumsum[f], int(mRNA_metadata_t[mRNA_all_list[f][p],0]-1)] = Merfish_CM[cell_to_map+cell_cumsum[f], int(mRNA_metadata_t[mRNA_all_list[f][p],0]-1)] + 1\n",
    "                Merfish_CM_list[f][cell_to_map, int(mRNA_metadata_t[mRNA_all_list[f][p],0]-1)] = Merfish_CM_list[f][cell_to_map, int(mRNA_metadata_t[mRNA_all_list[f][p],0]-1)] + 1\n",
    "\n",
    "\n",
    "                a = np.array([iter1,f,p,cell_to_map]) #added \n",
    "                asgn_stats = np.vstack((asgn_stats,a)) #added \n",
    "        '''nx31\n",
    "        else:\n",
    "            plt.figure(fig4)#,dpi=dpi_set,figsize=(12,12))\n",
    "            plt.plot(xc,yc, '.', markersize=4, c='black')#, alpha=0.5)\n",
    "        '''\n",
    "    '''nx31\n",
    "    plt.figure(fig4)#,dpi=dpi_set,figsize=(12,12))\n",
    "    plt.title(\"Iter_\"+str(iter1)+\"_\"+\"Fov_\"+str(f))\n",
    "    plt.axis('off')\n",
    "    file_name = os.path.join(path_python+'/figures_python/mRNA_assigned_seg_all_Iter='+str(iter1)+'_Fov='+str(f)+'.png')\n",
    "    plt.savefig(file_name)\n",
    "    '''\n",
    "    asgn_stats = np.delete(asgn_stats, (0), axis=0)\n",
    "    #asgn_stats_list.append(asgn_stats)\n",
    "    #end of all fov\n",
    "    return asgn_stats, mRNA_asgn_counter#, mRNA_assign_list # added mRNA_assign_list 1st July 2020\n",
    "     \n",
    "\n",
    "for iter1 in range(1,iterations):\n",
    "\n",
    "        #mRNA_assign_listoflists = []#added on 2nd June 2020\n",
    "        ##commneted the above on 4th July 2020\n",
    "        print('Iteration is: ', str(iter1))\n",
    "        #asgn_stats_list = []\n",
    "        Merfish_CM = np.zeros((1,Merfish_genes)) #will have all the Count matrices consolidated\n",
    "        for f in range(num_fov):\n",
    "            Merfish_CM = np.concatenate((Merfish_CM,Merfish_CM_list[f]),axis=0) #20th june 2020\n",
    "\n",
    "\n",
    "        Merfish_CM = np.delete(Merfish_CM, (0), axis=0)\n",
    "\n",
    "        #remove rows that have 0 rowsum\n",
    "\n",
    "        row_sums_CM = Merfish_CM.sum(axis=1,keepdims=True)\n",
    "        rows_remove = np.where(row_sums_CM==0)\n",
    "        Merfish_CM = np.delete(Merfish_CM,(rows_remove),axis=0)\n",
    "\n",
    "        #added 20th June 2020\n",
    "        cell_cumsum = np.cumsum(num_cells_list)\n",
    "        cell_cumsum = np.hstack((0,cell_cumsum))\n",
    "\n",
    "        for i in range(len(rows_remove[0])):\n",
    "            print('i is', str(i))\n",
    "            for j in range(i+1):\n",
    "                print('j is', str(j))\n",
    "                if ((j+1) < (len(cell_cumsum))):\n",
    "                    if (rows_remove[0][i] < cell_cumsum[j+1]):\n",
    "                        cell_cumsum[j+1] = cell_cumsum[j+1] - 1\n",
    "                        print(cell_cumsum)\n",
    "                continue\n",
    "            continue\n",
    "\n",
    "        #end add\n",
    "\n",
    "\n",
    "        M_upd = zscore(Merfish_CM,axis=1)\n",
    "\n",
    "        num_cells = M_upd.shape[0]\n",
    "        num_genes = M_upd.shape[1]\n",
    "        #save M_upd\n",
    "        df_data = pd.DataFrame(M_upd)\n",
    "        df_data.to_csv(os.path.join(path_python + '/count_data/count_matrix_Mupd_fovall_Iter_'+str(iter1)+'.csv'), header=None, index=None)\n",
    "\n",
    "        #save count matrix \n",
    "        df_data = pd.DataFrame(Merfish_CM)\n",
    "        df_data.to_csv(os.path.join(path_python + '/count_data/count_matrix_Merfish_fovall_Iter_'+str(iter1)+'.csv'), header=None, index=None)\n",
    "\n",
    "\n",
    "        #perform clustering for the 'sure' pixels\n",
    "        if(bayes_MM):\n",
    "            Mod_dpgmm=BayesianGaussianMixture(n_components=num_comp, covariance_type='full',weight_concentration_prior_type='dirichlet_process').fit(M_upd)\n",
    "            cluster_asgn = Mod_dpgmm.predict(M_upd)\n",
    "        else:\n",
    "            communities, graph, Q = pg.cluster(M_upd)#, k=num_comp, min_cluster_size=1)\n",
    "            cluster_asgn = communities\n",
    "\n",
    "        df_clusterlabel = pd.DataFrame(cluster_asgn)\n",
    "        df_clusterlabel.to_csv(os.path.join(path_python + '/count_data/cluster_labels_postreasgn_Iter_'+str(iter1)+'.csv'), header=None,\n",
    "                   index=None)\n",
    "\n",
    "        X_2d = tsne.fit_transform(M_upd)\n",
    "        df_data = pd.DataFrame(X_2d)\n",
    "        df_data.to_csv(os.path.join(path_python + '/count_data/tSNE_embedding_post.csv'), header=None, index=None)\n",
    "\n",
    "        unique, counts = np.unique(cluster_asgn, return_counts=True)\n",
    "        #t-SNE plots\n",
    "        target_ids = range(len(unique))\n",
    "        plt.figure(figsize=(8,8),frameon=False,dpi=dpi_set)\n",
    "        plt.axis('off')\n",
    "        #colors = 'r', 'g', 'b', 'c', 'm', 'y', 'k', 'w', 'orange', 'purple'\n",
    "        for i,  label in zip(target_ids, unique):\n",
    "            plt.scatter(X_2d[ cluster_asgn==(i), 0], X_2d[ cluster_asgn==(i), 1], color=cm[i], label=label,s=s_size)\n",
    "        plt.legend()\n",
    "        plt.title('t-SNE')\n",
    "        plt.show()\n",
    "        file_name = os.path.join(path_python+'/figures_python/tSNE_labels_postreasgn_newcoord_Iter_'+str(iter1)+'.png')\n",
    "        plt.savefig(file_name,dpi=dpi_set)\n",
    "\n",
    "        #mRNA_asgn_counter = np.zeros([num_fov,1])\n",
    "        #####added 20th June 2020\n",
    "        ##### recomputing the cluster moments based on the revised Merfish CMq\n",
    "        #####\n",
    "        mean_mat = [] #1 x genes per entry. K x genes overall\n",
    "        cov_list = [] # genes x genes per entry. K genes x genes overall\n",
    "\n",
    "        for i in range(len(unique)): #target_ids:\n",
    "            print(i)\n",
    "            #rows = np.where(cluster_asgn==unique[i])\n",
    "            rows = np.where(cluster_asgn==unique[i])\n",
    "            rows = np.array(rows).flatten()\n",
    "\n",
    "            if (rows.shape[0] > 1):\n",
    "                mean_mat.append(np.mean(M_upd[rows, :], axis=0))\n",
    "                cov_mat = np.cov(M_upd[rows, 0:num_genes].T)\n",
    "                cov_mat = np.dot(cov_mat,cov_mat.T)\n",
    "                np.fill_diagonal(cov_mat, np.diag(cov_mat)+0.01)\n",
    "                if (givens):\n",
    "                    (Q, R) = givens_rotation(cov_mat)\n",
    "                    cov_list.append(R)\n",
    "                else:\n",
    "                    cov_list.append(cov_mat)\n",
    "                \n",
    "        ##add end\n",
    "\n",
    "        t3 = time.time()\n",
    "        results = Parallel(n_jobs=num_cores)(delayed(processItern)(f) for f in inputs)\n",
    "        t4 = time.time()\n",
    "        print('Time taken to perform Iteration',str(iter1),'is:', str(t4-t3),'secs')\n",
    "        print('Time taken to perform all Iteration:',str(iter1),'is:', str((t4-t3)/60),'mins')\n",
    "        asgn_stats_list = [item[0] for item in results] #results#[item[0] for item in results]    \n",
    "        assigned_percent = [item[1] for item in results] \n",
    "        \n",
    "        \n",
    "        ## added 2nd July 2020\n",
    "        #mRNA_assign_listoflists = [item[2] for item in results]\n",
    "        #mRNA_assign_listoflists.append(item[2] for item in results) ## added 4th July 2020\n",
    "        #mRNA_assign_list = []        \n",
    "        ## added end\n",
    "        \n",
    "        asgn_stats_listoflists.append(asgn_stats_list)\n",
    "        ## added 20th June 2020\n",
    "\n",
    "        ###  added\n",
    "        ### 17th Nov 2019\n",
    "        #s=np.array(mRNA_asgn_counter_list).flatten()\n",
    "\n",
    "        #11th Dec 2019 -  added to remove ss just accessing FOV 0\n",
    "        ss= np.array(np.where(np.array(assigned_percent).flatten()!=0)).flatten()\n",
    "        #ss = np.array(np.where(assigned_percent!=0)).flatten() #new set of FOVs\n",
    "        #s = s[ss]\n",
    "        #mRNA_asgn_counter_list = s\n",
    "        print('within processItern loop')\n",
    "        print('ss is ',str(ss))\n",
    "        num_f = len(ss)\n",
    "        print('num_f is ',str(num_f))\n",
    "        print(ss)\n",
    "\n",
    "        #num_f = num_fov - 1\n",
    "        #num_fov = len(ss)\n",
    "        ##end add\n",
    "\n",
    "        print(\"assigned_percent \",str(assigned_percent))\n",
    "        '''\n",
    "        if (not np.any(assigned_percent)):\n",
    "            inputs=ss\n",
    "            break\n",
    "        if len(ss)!=0:\n",
    "            for kk in range(0,num_fov):\n",
    "                print('% of mRNA assigned over total dangling mRNAs in FOV ', str(ss[kk]),'is ', \n",
    "                  str(int(assigned_percent[ss[kk]])/len(mRNA_assign_list[ss[kk]])))\n",
    "                print('% of mRNA assigned over total mRNAs in FOV ', str(ss[kk]),'is ', \n",
    "                  str(int(assigned_percent[ss[kk]])/len(mRNA_all_list[ss[kk]])))\n",
    "        '''\n",
    "inputs = ss\n",
    "print('Overall runtime is:', str((t4-t1)/60),'mins')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#final Merfish_CM updated\n",
    "\n",
    "Merfish_CM = np.zeros((1,Merfish_genes)) #will have all the Count matrices consolidated\n",
    "for f in range(num_fov):\n",
    "    Merfish_CM = np.concatenate((Merfish_CM,Merfish_CM_list[f]),axis=0) #20th june 2020\n",
    "\n",
    "\n",
    "Merfish_CM = np.delete(Merfish_CM, (0), axis=0)\n",
    "\n",
    "#remove rows that have 0 rowsum\n",
    "\n",
    "row_sums_CM = Merfish_CM.sum(axis=1,keepdims=True)\n",
    "rows_remove = np.where(row_sums_CM==0)\n",
    "Merfish_CM = np.delete(Merfish_CM,(rows_remove),axis=0)\n",
    "\n",
    "#added 20th June 2020\n",
    "'''cell_cumsum = np.cumsum(num_cells_list)\n",
    "cell_cumsum = np.hstack((0,cell_cumsum))\n",
    "\n",
    "for i in range(len(rows_remove[0])):\n",
    "    print('i is', str(i))\n",
    "    for j in range(i+1):\n",
    "        print('j is', str(j))\n",
    "        if ((j+1) < (len(cell_cumsum))):\n",
    "            if (rows_remove[0][i] < cell_cumsum[j+1]):\n",
    "                cell_cumsum[j+1] = cell_cumsum[j+1] - 1\n",
    "                print(cell_cumsum)\n",
    "        continue\n",
    "    continue\n",
    "\n",
    "#end add\n",
    "'''\n",
    "\n",
    "M_upd = zscore(Merfish_CM,axis=1)\n",
    "\n",
    "num_cells = M_upd.shape[0]\n",
    "#num_genes = M_upd.shape[1]\n",
    "num_genes = Merfish_genes #nx31\n",
    "#save M_upd \n",
    "df_data = pd.DataFrame(M_upd)\n",
    "df_data.to_csv(os.path.join(path_python + '/count_data/count_matrix_Mupd_fovall_Iter_'+str(iterations)+'.csv'), header=None, index=None)\n",
    "\n",
    "#save count matrix \n",
    "df_data = pd.DataFrame(Merfish_CM)\n",
    "df_data.to_csv(os.path.join(path_python + '/count_data/count_matrix_Merfish_fovall_Iter_'+str(iterations)+'.csv'), header=None, index=None)\n",
    "\n",
    "#perform clustering for the 'sure' pixels\n",
    "if(bayes_MM):\n",
    "    Mod_dpgmm=BayesianGaussianMixture(n_components=num_comp, covariance_type='full',weight_concentration_prior_type='dirichlet_process').fit(M_upd)\n",
    "    cluster_asgn = Mod_dpgmm.predict(M_upd)\n",
    "else:\n",
    "    communities, graph, Q = pg.cluster(M_upd)#, k=num_comp, min_cluster_size=1)\n",
    "    cluster_asgn = communities\n",
    "\n",
    "df_clusterlabel = pd.DataFrame(cluster_asgn)\n",
    "df_clusterlabel.to_csv(os.path.join(path_python + '/count_data/cluster_labels_postreasgn_Iter_'+str(iterations)+'.csv'), header=None,\n",
    "       index=None)\n",
    "\n",
    "X_2d = tsne.fit_transform(M_upd)\n",
    "df_data = pd.DataFrame(X_2d)\n",
    "df_data.to_csv(os.path.join(path_python + '/count_data/tSNE_embedding_post_'+str(iterations)+'.csv'), header=None, index=None)\n",
    "\n",
    "unique, counts = np.unique(cluster_asgn, return_counts=True)\n",
    "#t-SNE plots\n",
    "target_ids = range(len(unique))\n",
    "plt.figure(figsize=(10,10),frameon=False,dpi=dpi_set)\n",
    "plt.axis('off')\n",
    "#colors = 'r', 'g', 'b', 'c', 'm', 'y', 'k', 'w', 'orange', 'purple'\n",
    "for i,  label in zip(target_ids, unique):\n",
    "    plt.scatter(X_2d[ cluster_asgn==(i), 0], X_2d[ cluster_asgn==(i), 1], color=cm[i], label=label,s=s_size)\n",
    "plt.legend()\n",
    "plt.title('t-SNE')\n",
    "plt.show()\n",
    "file_name = os.path.join(path_python+'/figures_python/tSNE_labels_postreasgn_newcoord_Iter_'+str(iterations)+'.png')\n",
    "plt.savefig(file_name,dpi=dpi_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################################################\n",
    "## parallelise the overall merging of mRNA assignments per FOV ####\n",
    "###################################################################\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def processOverall(f):\n",
    "\n",
    "    #asgn_stats = np.array([iter1,0,0,0]) #added \n",
    "\n",
    "    if(fig2p==1):\n",
    "        fig2 = (f+1)*200+(f+1)\n",
    "    #fig4 = (iter1+f+1)*4+(iter1+f+1)\n",
    "    fig4 = (iter1+f+1)*4+f\n",
    "\n",
    "    #fig2 = (f+1)*200+(f+1)\n",
    "    #fig4 = (f+1)*400+(f+1)\n",
    "\n",
    "    plt.figure(fig4,dpi=dpi_set,figsize=(12,12))\n",
    "    plt.title(str(f))\n",
    "    plt.axis('off')\n",
    "\n",
    "    print('Fov is: ', str(f))\n",
    "\n",
    "    num_cells = cell_metadata_t_list[f].shape[0]\n",
    "\n",
    "\n",
    "    #perspective projection\n",
    "    if z==1:\n",
    "        z_x = 6\n",
    "        z_y = 7\n",
    "    else:\n",
    "        z_x = 6 + 2**(z-1)\n",
    "        z_y = 7 + 2**(z-1)\n",
    "\n",
    "    rw_x_boundary = cell_metadata_t_list[f][:,z_x]\n",
    "    rw_y_boundary = cell_metadata_t_list[f][:,z_y]\n",
    "\n",
    "    CoM_stage_coords = np.zeros([num_cells, 2], dtype=np.float)\n",
    "    CoM_image_coords = np.zeros([num_cells, 2], dtype=np.float)\n",
    "\n",
    "    #fig1 = (f+1)*100+(f+1)\n",
    "\n",
    "\n",
    "    print('plotting cells')\n",
    "    for i in range(num_cells): #i will be i+2th row in the xls\n",
    "\n",
    "        #print(\"Cell= \" + str(i))\n",
    "\n",
    "        if(is_nan(rw_x_boundary[i]) | is_nan(rw_y_boundary[i])): #entries were empty: no segmentation found\n",
    "            print(\"skip cell \", str(i))\n",
    "        else:\n",
    "            m = rw_x_boundary[i].split(';')\n",
    "            n = rw_y_boundary[i].split(';')\n",
    "            num_coord = m.__len__()\n",
    "\n",
    "            # this can be done smarter. for now I cast every string float to float and place into a list\n",
    "            m_list = []\n",
    "            n_list = []\n",
    "            for j in range(num_coord-1):\n",
    "                if( (float(m[j])!=float(m[j]))  or (float(n[j])!=float(n[j]))):\n",
    "                    print(\"Coord NaN is at = \" + str(j))\n",
    "                    #continue\n",
    "                else:\n",
    "                    m_list.append(float(m[j]))\n",
    "                    n_list.append(float(n[j]))\n",
    "\n",
    "            #to plot the stage coords\n",
    "            #for k in np.arange(0, m_list.__len__()-1, 2):\n",
    "                #print(k)\n",
    "                #plt.figure(fig1,figsize=(10,10),dpi=dpi_set)\n",
    "            #    connectpoints(m_list, n_list, k, k + 1)\n",
    "\n",
    "            x_coord = np.array(m_list)  # (256,)\n",
    "            y_coord = np.array(n_list)  # (256,)\n",
    "            f1 = np.vstack((x_coord, y_coord))  # (2, 256)\n",
    "\n",
    "            cX = sum(f1[0, :]) / f1.shape[1]\n",
    "            cY = sum(f1[1, :]) / f1.shape[1]\n",
    "\n",
    "            CoM_stage_coords[i, 0] = cX\n",
    "            CoM_stage_coords[i, 1] = cY\n",
    "            im_list = []\n",
    "            in_list = []\n",
    "            for s in range(m_list.__len__()):\n",
    "                a = np.array([[f1[0, s], f1[1, s]]], dtype='float32')\n",
    "                a = np.array([a])\n",
    "                pointsOut = cv2.perspectiveTransform(a, h)\n",
    "                im_list.append(pointsOut[0,0,0])\n",
    "                in_list.append(pointsOut[0,0,1])\n",
    "\n",
    "            for k in np.arange(0, im_list.__len__()-1, 2):\n",
    "\n",
    "                plt.figure(fig4,figsize=(10,10),dpi=dpi_set)\n",
    "                connectpoints(im_list, in_list, k, k + 1)\n",
    "\n",
    "            ix_coord = np.array(im_list)  # (256,)\n",
    "            iy_coord = np.array(in_list)  # (256,)\n",
    "            f_image = np.vstack((ix_coord, iy_coord))  # (2, 256)\n",
    "\n",
    "            cX_image = sum(f_image[0, :]) / f_image.shape[1]\n",
    "            cY_image = sum(f_image[1, :]) / f_image.shape[1]\n",
    "\n",
    "            CoM_image_coords[i,0] = cX_image\n",
    "            CoM_image_coords[i,1] = cY_image\n",
    "\n",
    "\n",
    "            if(fig2p==1):\n",
    "                plt.figure(fig2) #  ,dpi=dpi_set,figsize=(10,10))\n",
    "                axes = plt.gca()\n",
    "                #axes.add_patch(Polygon(np.transpose(f_image),closed=True, facecolor=colours[cluster_asgn[i]]))\n",
    "                axes.add_patch(Polygon(np.transpose(f_image),closed=True, facecolor=cm.colors[cluster_asgn[i]]))\n",
    "\n",
    "\n",
    "        ## plot mRNA inside the cell\n",
    "        ##\n",
    "        feat_id = cell_metadata_t_list[f][i,1]\n",
    "        feat_rows = np.where((mRNA_metadata_t[:,1]==f) & (mRNA_metadata_t[:,9]==feat_id) & (mRNA_metadata_t[:,10]==1)) #fov id, which cell and in/out of cell\n",
    "        #feat_rows = np.where((mRNA_metadata_t[:, 1] == 0) & (mRNA_metadata_t[:, 9] == feat_id))\n",
    "        feat_rows = np.array(feat_rows).flatten()\n",
    "        num_rows = feat_rows.shape[0]\n",
    "        for j in range(num_rows):\n",
    "            x_stage_coord = mRNA_metadata_t[feat_rows[j],3]\n",
    "            y_stage_coord = mRNA_metadata_t[feat_rows[j],4]\n",
    "            a = np.array([[x_stage_coord, y_stage_coord]], dtype='float32')\n",
    "            a = np.array([a])\n",
    "\n",
    "            #perspective proj\n",
    "            pointsOut = cv2.perspectiveTransform(a, h)\n",
    "\n",
    "            x_image_coord = pointsOut[0,0,0]\n",
    "            y_image_coord = pointsOut[0,0,1]\n",
    "\n",
    "            #plt.plot(x_image_coord, y_image_coord,'o',markersize=2, c=colours[cluster_asgn[i]],alpha=0.5)\n",
    "            if ((i+cell_cumsum[f]) < M_upd.shape[0]):\n",
    "                plt.plot(x_image_coord, y_image_coord,'.',markersize=4, color=cm[cluster_asgn[i+cell_cumsum[f]]])\n",
    "\n",
    "\n",
    "    mRNA_im_coords = mRNA_coords_list[f]  #added # all mRNA are always the same per fov. so this is ok, no need to keep track of the iteration\n",
    "\n",
    "    plt.figure(fig4,dpi=dpi_set,figsize=(12,12))\n",
    "    for itert in range(iterations):\n",
    "        print(\"within final mergen iteration is:\", str(itert))\n",
    "        if (f < len(asgn_stats_listoflists[itert])):\n",
    "            to_plot_mRNA = asgn_stats_listoflists[itert][f]\n",
    "            len_tp = len(to_plot_mRNA)\n",
    "            #if (len_tp > 0):\n",
    "            if (isinstance(to_plot_mRNA[0], np.ndarray)): \n",
    "                print('len_tp', len_tp)\n",
    "                #nx31 print('to_plot_mRNA', to_plot_mRNA)\n",
    "\n",
    "                for l in range(len_tp):\n",
    "\n",
    "                    pp = to_plot_mRNA[l,2]\n",
    "                    cell_pp = to_plot_mRNA[l,3]\n",
    "\n",
    "                    if(pp < mRNA_im_coords.shape[0]):\n",
    "                        xc = mRNA_im_coords[pp, 0]\n",
    "                        yc = mRNA_im_coords[pp, 1]\n",
    "\n",
    "                        if(cell_pp < CoM_image_coords.shape[0]):\n",
    "                            if ((cell_pp+cell_cumsum[f]) < M_upd.shape[0]):\n",
    "                                plt.plot(xc, yc,'x',markersize=4, c=cm[cluster_asgn[cell_pp+cell_cumsum[f]]],alpha=0.5)\n",
    "\n",
    "                                x_temp = [CoM_image_coords[cell_pp,0],xc]\n",
    "                                y_temp = [CoM_image_coords[cell_pp,1],yc]\n",
    "\n",
    "                                plt.plot(x_temp, y_temp, linewidth=1, c=cm[cluster_asgn[cell_pp+cell_cumsum[f]]])\n",
    "\n",
    "\n",
    "    plt.figure(fig4)\n",
    "    plt.title(\"Final_plot_Fov_\"+str(f))\n",
    "    plt.axis('off')\n",
    "    file_name = os.path.join(path_python+'/figures_python/overall_mRNA_assigned_seg_all_Fov='+str(f)+'.png')\n",
    "    plt.savefig(file_name,dpi=dpi_set)\n",
    "\n",
    "\n",
    "t5 = time.time()\n",
    "inputs = range(num_fov)\n",
    "\n",
    "results = Parallel(n_jobs=num_cores)(delayed(processOverall)(f) for f in inputs)\n",
    "t6 = time.time()\n",
    "print('Time taken to perform overall merge is',str(t6-t5),'secs')\n",
    "print('Time taken to perform overall merge is',str((t6-t5)/60),'mins')\n",
    "\n",
    "\n",
    "print('Overall runtime is:', str((t6-t1)/60),'mins')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
